br/pkg/restore/import.go

```diff
@@ -866,6 +850,284 @@ func (importer *FileImporter) downloadRawKVSST(
+	if err != nil {
+		return nil, import_sstpb.SSTMeta{}, err
+	}
+
+	req := &import_sstpb.DownloadRequest{
+		Sst:            *sstMeta,
+		StorageBackend: importer.backend,
+		Name:           file.GetName(),
+		RewriteRule:    rule,
+		CipherInfo:     cipher,
+		StorageCacheId: importer.cacheKey,
+		// For the older version of TiDB, the request type will  be default to `import_sstpb.RequestType_Legacy`
+		RequestType: import_sstpb.DownloadRequestType_Keyspace,
+		Context: &kvrpcpb.Context{
+			ResourceControlContext: &kvrpcpb.ResourceControlContext{
+				ResourceGroupName: "", // TODO,
+			},
+			RequestSource: kvutil.BuildRequestSource(true, kv.InternalTxnBR, kvutil.ExplicitTypeBR),
+		},
+	}
+	return req, *sstMeta, nil
+}
+
+func (importer *FileImporter) downloadSSTV2(
+	ctx context.Context,
+	regionInfo *split.RegionInfo,
+	files []*backuppb.File,
+	rewriteRules *RewriteRules,
+	cipher *backuppb.CipherInfo,
+	apiVersion kvrpcpb.APIVersion,
+) ([]*import_sstpb.SSTMeta, error) {
+	var mu sync.Mutex
+	downloadMetas := make([]*import_sstpb.SSTMeta, 0, len(files))
+	downloadMetasMap := make(map[string]import_sstpb.SSTMeta)
+	resultMetasMap := make(map[string]*import_sstpb.SSTMeta)
+	downloadReqsMap := make(map[string]*import_sstpb.DownloadRequest)
+	for _, file := range files {
+		req, sstMeta, err := importer.buildDownloadRequest(file, rewriteRules, regionInfo, cipher)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		downloadMetasMap[file.Name] = sstMeta
+		downloadReqsMap[file.Name] = req
+	}
+
+	eg, ectx := errgroup.WithContext(ctx)
+	for _, p := range regionInfo.Region.GetPeers() {
+		peer := p
+		eg.Go(func() error {
+			importer.storeWorkerPoolRWLock.Lock()
+			workerCh, ok := importer.storeWorkerPoolMap[peer.GetStoreId()]
+			// handle the case that the store is new-scaled in the cluster
+			if !ok {
+				workerCh = utils.BuildWorkerTokenChannel(importer.concurrencyPerStore)
+				importer.storeWorkerPoolMap[peer.GetStoreId()] = workerCh
+			}
+			importer.storeWorkerPoolRWLock.Unlock()
+			defer func() {
+				workerCh <- struct{}{}
+			}()
+			<-workerCh
+			for _, file := range files {
+				req, ok := downloadReqsMap[file.Name]
+				if !ok {
+					return errors.New("not found file key for download request")
+				}
+				var err error
+				var resp *import_sstpb.DownloadResponse
+				err = utils.WithRetry(ctx, func() error {
+					dctx, cancel := context.WithTimeout(ectx, gRPCTimeOut)
+					resp, err = importer.importClient.DownloadSST(dctx, peer.GetStoreId(), req)
+					cancel()
+					if err != nil {
+						return err
+					}
+					return nil
+				}, utils.NewDownloadSSTBackoffer())
+				if err != nil {
+					return errors.Trace(err)
+				}
+				if resp.GetError() != nil {
+					return errors.Annotate(berrors.ErrKVDownloadFailed, resp.GetError().GetMessage())
+				}
+				if resp.GetIsEmpty() {
+					return errors.Trace(berrors.ErrKVRangeIsEmpty)
```

Comment:
Maybe also provide the range of the file for debugging.

Reply:
we had log outside
```
log.Warn("download file skipped",
	logutil.Files(files),
	logutil.Region(info.Region),
	logutil.Key("startKey", startKey),
	logutil.Key("endKey", endKey),
	logutil.Key("file-simple-start", files[0].StartKey),
	logutil.Key("file-simple-end", files[0].EndKey),
```

Reply:
It seems the file's key range won't be surely print? We only print the first file there. Even in most case there should be only one file, querying log is usually harder than the direct error message.

Reply:
`ErrKVRangeIsEmpty` is always a `retryable error` and never break the restoration. it appears somehow doesn't really means error. I think is kind of feature rather than error.